common/src/main/java/org/opensearch/ml/common/connector/ConnectorClientConfig.java

```diff
@@ -26,45 +29,83 @@ public class ConnectorClientConfig implements ToXContentObject, Writeable {
+    public static final String RETRY_TIMEOUT_SECONDS_FIELD = "retry_timeout_seconds";
+    public static final String MAX_RETRY_TIMES_FIELD = "max_retry_times";
+    public static final String RETRY_BACKOFF_POLICY_FIELD = "retry_backoff_policy";
 
     public static final Integer MAX_CONNECTION_DEFAULT_VALUE = Integer.valueOf(30);
     public static final Integer CONNECTION_TIMEOUT_DEFAULT_VALUE = Integer.valueOf(30000);
     public static final Integer READ_TIMEOUT_DEFAULT_VALUE = Integer.valueOf(30000);
+    public static final Integer RETRY_BACKOFF_MILLIS_DEFAULT_VALUE = 200;
+    public static final Integer RETRY_TIMEOUT_SECONDS_DEFAULT_VALUE = 30;
+    public static final Integer MAX_RETRY_TIMES_DEFAULT_VALUE = -1;
+    public static final String RETRY_BACKOFF_POLICY_DEFAULT_VALUE = "constant";
+    public static final Version MINIMAL_SUPPORTED_VERSION_FOR_RETRY = Version.V_2_15_0;
+    public static final List<String> ALL_SUPPORTED_BACKOFF_POLICY = List.of("constant", "exponential_equal_jitter", "exponential_full_jitter");
 
     private Integer maxConnections;
     private Integer connectionTimeout;
     private Integer readTimeout;
+    private Integer retryBackoffMillis;
+    private Integer retryTimeoutSeconds;
+    private Integer maxRetryTimes;
+    private String retryBackoffPolicy;
 
     @Builder(toBuilder = true)
     public ConnectorClientConfig(
         Integer maxConnections,
         Integer connectionTimeout,
-        Integer readTimeout
+        Integer readTimeout,
+        Integer retryBackoffMillis,
+        Integer retryTimeoutSeconds,
+        Integer maxRetryTimes,
+        String retryBackoffPolicy
     ) {
         this.maxConnections = maxConnections;
         this.connectionTimeout = connectionTimeout;
         this.readTimeout = readTimeout;
-
+        this.retryBackoffMillis = retryBackoffMillis;
+        this.retryTimeoutSeconds = retryTimeoutSeconds;
+        this.maxRetryTimes = maxRetryTimes;
+        this.retryBackoffPolicy = retryBackoffPolicy;
     }
 
     public ConnectorClientConfig(StreamInput input) throws IOException {
+        Version streamInputVersion = input.getVersion();
         this.maxConnections = input.readOptionalInt();
         this.connectionTimeout = input.readOptionalInt();
         this.readTimeout = input.readOptionalInt();
+        if(streamInputVersion.onOrAfter(MINIMAL_SUPPORTED_VERSION_FOR_RETRY)) {
+            this.retryBackoffMillis = input.readOptionalInt();
+            this.retryTimeoutSeconds = input.readOptionalInt();
+            this.maxRetryTimes = input.readOptionalInt();
+            this.retryBackoffPolicy = input.readOptionalString();
+        }
     }
 
     public ConnectorClientConfig() {
         this.maxConnections = MAX_CONNECTION_DEFAULT_VALUE;
         this.connectionTimeout = CONNECTION_TIMEOUT_DEFAULT_VALUE;
         this.readTimeout = READ_TIMEOUT_DEFAULT_VALUE;
+        this.retryBackoffMillis = RETRY_BACKOFF_MILLIS_DEFAULT_VALUE;
```

Comment:
So no matter user configure it or not, we will always add retry for all connectors? That's a breaking change for the old versions.

Reply:
+1, can we add a flag here to disable the retry by default so BWC will not break?

Reply:
This is an enhancements to improve the user experience, so I believe we should open it by default. For bwc, I tested restart upgrade and rolling upgrade, this PR doesn't break anything

Reply:
Agree, this brings benefit to customers, we should enable this by default.

Reply:
This is the lowest level function that many other application depends on like neural search. They have implemented retry 3 times by default. This PR enforces every dependency to retry at least 3 times by default. In the neural search they would retry 9 times when throttled. It's breaking the BWC behavior at least for retry times.

Reply:
In throttling, does customer really wants retry 3 times more than their setups?

Reply:
Think about this, user has some code which on top of ml-commons, they have implemented some retry , like retry 5 times. Now we add another retry in ml-commons , like 3 times, so when this feature out, the  user will see 15 times retry , right ? That will be too much for them. Let's have a setting to control enable/disable this retry, or retry times.